File: blockchain/src/chain.py
"""
chain.py — Python <-> Rust bridge for the append-only verifiable ledger
(blockchain/, crate `sombra-chain`).

WHAT THIS IS. A private, single-node, append-only verifiable ledger. There
is one writer (this bot's host), no peers, no networking, no mining and no
consensus — so it is deliberately NOT described as a decentralized
blockchain anywhere in this codebase or in any user-facing Thai text.

WHAT IT IS NOT. Not a cryptocurrency, not a payment rail, not a second
source of truth for anybody's balance. `bot.db` remains the only place a
balance lives, and nothing in this module or in the Rust crate ever writes
to `wallets`, `wallet_transactions`, `debt_entries` or `expenses`.

HOW IT INTEGRATES — ANCHOR (PULL), NOT DUAL-WRITE. wallet.py,
debt_ledger.py and expense.py are not modified at all. The Rust binary
*reads* rows those modules have already committed and appends them to the
chain. The task brief's three consistency hazards are therefore structurally
impossible rather than merely handled:
  - "balance and ledger disagree": no balance is ever computed here;
  - "chain committed, database failed": the chain can only record rows the
    database already committed;
  - "database committed, chain failed": the chain lags and the next anchor
    run catches up, because every anchor is keyed by a deterministic
    `tx_ref` with a UNIQUE index and is therefore idempotent.
Recovery after a crash, a kill -9, or a week with the binary missing is the
same single action: run `anchor` again.

INTEGRATION STYLE — CLI over subprocess. Chosen over PyO3 bindings (which
would put a Rust toolchain in the critical path of every `pip install`) and
over a local service (a second process to supervise for something that runs
a few times an hour). The binary is also runnable by hand during debugging.

DEGRADED MODE IS A FIRST-CLASS PATH. If the binary is missing, not
executable, times out, crashes, or returns unparseable output, every
function here returns an OpResult-shaped failure and the bot carries on
exactly as before. This module raises nothing; the Telegram bot must never
go down because a ledger component is unhappy. That matters especially on
first deploy, when the Rust binary may not have been built yet.

Design constraints (matches security.py / wallet.py / expense.py):
- Standard library only (asyncio, json, logging, os, shutil, subprocess),
  plus `from security import DB_PATH` -- exactly what every other data-layer
  module here does. Hard-coding a second "bot.db" literal would mean that
  the day someone repoints security.DB_PATH, the ledger would silently keep
  anchoring the old database while the bot wrote to the new one.
- Owns no tables: the Rust crate creates and owns `chain_blocks`,
  `chain_transactions` and `chain_meta` inside the existing `bot.db`.
- No LLM calls, no network.
"""

import os
import json
import shutil
import asyncio
import logging
import subprocess

from security import DB_PATH

logger = logging.getLogger("modbot.chain")

CHAIN_ENABLED = os.getenv("CHAIN_ENABLED", "true").lower() != "false"
CHAIN_BINARY_ENV = os.getenv("CHAIN_BINARY", "").strip()
CHAIN_ANCHOR_INTERVAL = int(os.getenv("CHAIN_ANCHOR_INTERVAL", "300"))
CHAIN_MAX_TX_PER_BLOCK = int(os.getenv("CHAIN_MAX_TX_PER_BLOCK", "500"))
CHAIN_TIMEOUT_SECONDS = float(os.getenv("CHAIN_TIMEOUT", "30"))
CHAIN_STARTUP_DELAY = int(os.getenv("CHAIN_STARTUP_DELAY", "20"))

_HERE = os.path.dirname(os.path.abspath(__file__))
_CANDIDATE_PATHS = (
    os.path.join(_HERE, "blockchain", "target", "release", "sombra-chain"),
    os.path.join(_HERE, "blockchain", "target", "debug", "sombra-chain"),
)


class ChainResult:
    """Same shape as wallet.py's / expense.py's OpResult so app.py handles a
    chain failure with the existing `if not result.ok: deny_text(...)`
    pattern instead of a second convention."""

    __slots__ = ("ok", "reason", "detail", "data")

    def __init__(self, ok, reason="", detail="", data=None):
        self.ok = ok
        self.reason = reason
        self.detail = detail
        self.data = data or {}

    def __repr__(self):
        return f"ChainResult(ok={self.ok!r}, reason={self.reason!r})"


_warned_about_binary_env = False


def find_binary():
    """Locates the compiled Rust binary, or returns None.

    Search order: an explicitly configured CHAIN_BINARY, then the release
    build, then the debug build, then PATH. The candidate paths are built
    from this file's own absolute location, not the process working
    directory, so they survive the bot being started from anywhere.

    Checked on every call rather than cached at import: a deploy that builds
    the binary after the worker has already started then heals on the next
    anchor tick instead of needing a restart.

    An explicitly configured CHAIN_BINARY is NOT silently fallen back on
    when it is unusable -- a wrong path in the environment is a
    misconfiguration to surface, not to paper over by quietly running some
    other binary. It is logged once so it does not repeat on every tick."""
    global _warned_about_binary_env
    if CHAIN_BINARY_ENV:
        if os.access(CHAIN_BINARY_ENV, os.X_OK):
            return CHAIN_BINARY_ENV
        if not _warned_about_binary_env:
            _warned_about_binary_env = True
            logger.warning(
                "CHAIN: CHAIN_BINARY is set to %r but that path is not an "
                "executable file -- the ledger stays unavailable rather than "
                "falling back to another binary", CHAIN_BINARY_ENV,
            )
        return None
    for path in _CANDIDATE_PATHS:
        if os.access(path, os.X_OK):
            return path
    return shutil.which("sombra-chain")


def is_available():
    return CHAIN_ENABLED and find_binary() is not None


def _decode(stdout, stderr, returncode):
    """The Rust CLI prints exactly one JSON object on stdout for both
    success and failure, so parse first and only fall back to the exit code
    if that fails."""
    text = (stdout or b"").decode("utf-8", errors="replace").strip()
    if not text:
        detail = (stderr or b"").decode("utf-8", errors="replace").strip()
        return ChainResult(False, reason="CHAIN_NO_OUTPUT",
                           detail=detail[:500] or f"exit code {returncode}")
    try:
        payload = json.loads(text)
    except json.JSONDecodeError:
        logger.warning("CHAIN: unparseable output from sombra-chain: %s", text[:300])
        return ChainResult(False, reason="CHAIN_BAD_OUTPUT", detail=text[:500])
    if not isinstance(payload, dict):
        return ChainResult(False, reason="CHAIN_BAD_OUTPUT", detail=text[:500])
    if payload.get("ok") is False and "error" in payload:
        return ChainResult(False, reason=str(payload.get("error")),
                           detail=str(payload.get("detail", ""))[:500], data=payload)
    return ChainResult(True, data=payload)


def _argv(binary, command, extra=None):
    argv = [binary, command, "--db", DB_PATH]
    if extra:
        argv.extend(extra)
    return argv


async def _run(command, extra=None, timeout=None):
    """Runs one CLI subcommand without blocking the event loop.

    Every failure mode -- binary missing, non-zero exit, timeout, OS-level
    spawn error -- comes back as a ChainResult. Nothing raises."""
    if not CHAIN_ENABLED:
        return ChainResult(False, reason="CHAIN_DISABLED")
    binary = find_binary()
    if not binary:
        return ChainResult(False, reason="CHAIN_UNAVAILABLE")

    argv = _argv(binary, command, extra)
    try:
        process = await asyncio.create_subprocess_exec(
            *argv,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
    except OSError:
        logger.exception("CHAIN: failed to start %s", binary)
        return ChainResult(False, reason="CHAIN_UNAVAILABLE")

    try:
        stdout, stderr = await asyncio.wait_for(
            process.communicate(), timeout=timeout or CHAIN_TIMEOUT_SECONDS
        )
    except asyncio.TimeoutError:
        # Kill rather than leave an orphan holding SQLite's write lock,
        # which would stall every /transfer and /debt_pay in the group.
        try:
            process.kill()
            await process.wait()
        except (ProcessLookupError, OSError):
            pass
        logger.warning("CHAIN: %s timed out after %.0fs", command, timeout or CHAIN_TIMEOUT_SECONDS)
        return ChainResult(False, reason="CHAIN_TIMEOUT")

    return _decode(stdout, stderr, process.returncode)


def _run_sync(command, extra=None, timeout=None):
    """Blocking variant, used only by chain_db_init() at process start --
    before the event loop is running, exactly like every other *_db_init()
    in this repo."""
    if not CHAIN_ENABLED:
        return ChainResult(False, reason="CHAIN_DISABLED")
    binary = find_binary()
    if not binary:
        return ChainResult(False, reason="CHAIN_UNAVAILABLE")
    try:
        completed = subprocess.run(
            _argv(binary, command, extra),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=timeout or CHAIN_TIMEOUT_SECONDS,
            check=False,
        )
    except subprocess.TimeoutExpired:
        return ChainResult(False, reason="CHAIN_TIMEOUT")
    except OSError:
        logger.exception("CHAIN: failed to start %s", binary)
        return ChainResult(False, reason="CHAIN_UNAVAILABLE")
    return _decode(completed.stdout, completed.stderr, completed.returncode)


# ---------------- Public API ----------------

def chain_db_init():
    """Creates the chain tables and the deterministic genesis block if they
    are absent. Idempotent and non-fatal: a missing or unbuilt binary logs a
    warning and the bot starts normally without the ledger."""
    if not CHAIN_ENABLED:
        logger.info("CHAIN DATABASE: disabled (CHAIN_ENABLED=false)")
        return ChainResult(False, reason="CHAIN_DISABLED")
    result = _run_sync("init")
    if result.ok:
        logger.info("CHAIN DATABASE: OK (genesis_created=%s)",
                    result.data.get("genesis_created"))
    else:
        logger.warning(
            "CHAIN DATABASE: unavailable (%s) -- the bot runs normally, but "
            "nothing is anchored until blockchain/ is built with "
            "`cargo build --release`", result.reason,
        )
    return result


async def anchor_now(max_tx=None):
    """Appends at most one block containing everything committed but not yet
    anchored. Safe to call concurrently with itself: the Rust side takes
    SQLite's write lock with BEGIN IMMEDIATE and the loser finds nothing
    left to do."""
    budget = CHAIN_MAX_TX_PER_BLOCK if max_tx is None else max_tx
    return await _run("anchor", ["--max-tx", str(budget)])


async def get_status():
    return await _run("status")


async def verify_chain(from_height=0):
    """Structural integrity of the chain itself: hashes, links, Merkle
    roots, duplicate transactions."""
    return await _run("verify", ["--from", str(max(0, int(from_height)))])


async def reconcile(limit=0):
    """Compares each anchored payload against the source row as it stands
    today -- catches an edit made straight against the database, which
    verify_chain() cannot see because the chain itself stays well formed.
    Rows still waiting to be anchored are skipped, so a backlog never looks
    like tampering."""
    return await _run("reconcile", ["--limit", str(max(0, int(limit)))])


async def get_block(height):
    return await _run("block", ["--height", str(int(height))])


async def find_transaction(needle):
    """Accepts a chain tx_ref ('wallet_tx:12', 'debt_entry:7:paid'), a
    'source:id' pair, or a bare wallet transaction id as shown by /history."""
    return await _run("tx", ["--ref", str(needle)])


# ---------------- Background anchoring ----------------

async def chain_background_loop():
    """Periodic anchoring, mirroring news.py's news_background_loop():
    sleep, work, swallow every exception, repeat.

    Deliberately a background loop rather than a hook on each wallet write.
    Anchoring inside a command handler would add a subprocess spawn to the
    latency of every /transfer and, worse, would want SQLite's write lock
    while wallet.py is still holding it. Lagging by one interval is the
    correct trade for an audit log."""
    if not CHAIN_ENABLED:
        return
    await asyncio.sleep(CHAIN_STARTUP_DELAY)
    while True:
        try:
            result = await anchor_now()
            if result.ok:
                anchored = result.data.get("anchored", 0)
                if anchored:
                    logger.info("CHAIN ANCHOR: %s transaction(s) into block %s",
                                anchored, result.data.get("height"))
            elif result.reason not in ("CHAIN_UNAVAILABLE", "CHAIN_DISABLED"):
                logger.warning("CHAIN ANCHOR FAILED: %s %s", result.reason, result.detail)
        except asyncio.CancelledError:
            raise
        except Exception:
            logger.exception("CHAIN BACKGROUND LOOP ERROR")
        await asyncio.sleep(CHAIN_ANCHOR_INTERVAL)

File: blockchain/src/chain.rs
//! Chain operations: bootstrap, anchor, status, lookup, reconcile.

use rusqlite::{params, Connection, TransactionBehavior};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::block::Block;
use crate::storage;
use crate::transaction::{
    self, ChainTransaction, SOURCE_DEBT, SOURCE_EXPENSE, SOURCE_WALLET, TX_DEBT_PAID,
    TX_EXPENSE_ADDED, TX_EXPENSE_DELETED, TX_EXPENSE_REVISED,
};
use crate::{ChainError, Result, DEFAULT_MAX_TX_PER_BLOCK};

pub fn now_epoch() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

/// Creates the schema and, if the chain is empty, writes the deterministic
/// genesis block. Idempotent: safe to run on every bot start.
pub fn init(conn: &mut Connection) -> Result<bool> {
    storage::init_schema(conn)?;
    if storage::height(conn)?.is_some() {
        return Ok(false);
    }
    let genesis = Block::genesis();
    let tx = conn.transaction_with_behavior(TransactionBehavior::Immediate)?;
    // Re-check inside the write lock: two processes may have raced here.
    let already: Option<i64> = {
        use rusqlite::OptionalExtension;
        tx.query_row("SELECT MAX(height) FROM chain_blocks", [], |row| {
            row.get::<_, Option<i64>>(0)
        })
        .optional()?
        .flatten()
    };
    if already.is_some() {
        tx.rollback()?;
        return Ok(false);
    }
    storage::insert_block(&tx, &genesis, now_epoch())?;
    tx.execute(
        "INSERT INTO chain_meta (key, value) VALUES ('chain_domain', ?1)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value",
        params![crate::CHAIN_DOMAIN],
    )?;
    tx.commit()?;
    Ok(true)
}

// ---------------- Anchoring ----------------

/// Reads source rows that the chain has not recorded yet.
///
/// Every predicate is a `NOT EXISTS` anti-join against `chain_transactions`,
/// so anchoring is exactly-once per state without a mutable cursor that
/// could drift, be reset, or skip a row that settled out of id order (a
/// pending deposit confirmed after a later transfer already committed).
fn collect_pending(conn: &Connection, budget: usize) -> Result<Vec<ChainTransaction>> {
    let mut out: Vec<ChainTransaction> = Vec::new();

    // --- wallet_transactions ---
    // Only terminal rows. wallet.py moves a row pending -> completed or
    // pending -> cancelled and never out again, so a terminal row's fields
    // are frozen and one anchor per row is correct and final.
    if storage::table_exists(conn, "wallet_transactions")? {
        let remaining = budget.saturating_sub(out.len());
        if remaining > 0 {
            let mut stmt = conn.prepare(
                "SELECT w.transaction_id, w.chat_id, w.user_id, w.type, w.amount_satang,
                        w.balance_before_satang, w.balance_after_satang, w.status,
                        w.reference_id, w.counterparty_user_id, w.created_at
                 FROM wallet_transactions w
                 WHERE w.status IN ('completed', 'cancelled')
                   AND NOT EXISTS (
                       SELECT 1 FROM chain_transactions c
                       WHERE c.source = 'wallet_tx' AND c.source_row_id = w.transaction_id)
                 ORDER BY w.transaction_id ASC
                 LIMIT ?1",
            )?;
            let rows = stmt.query_map(params![remaining as i64], |row| {
                Ok(transaction::from_wallet_row(
                    row.get(0)?,
                    row.get(1)?,
                    row.get(2)?,
                    &row.get::<_, String>(3)?,
                    row.get(4)?,
                    row.get(5)?,
                    row.get(6)?,
                    &row.get::<_, String>(7)?,
                    row.get(8)?,
                    row.get(9)?,
                    row.get(10)?,
                ))
            })?;
            for row in rows {
                out.push(row?);
            }
        }
    }

    // --- debt_entries: signed ---
    if storage::table_exists(conn, "debt_entries")? {
        let remaining = budget.saturating_sub(out.len());
        if remaining > 0 {
            let mut stmt = conn.prepare(
                "SELECT e.entry_id, e.chat_id, e.debtor_name, e.amount_satang,
                        e.item_description, e.entry_date, e.recorded_by, e.created_at
                 FROM debt_entries e
                 WHERE NOT EXISTS (
                       SELECT 1 FROM chain_transactions c
                       WHERE c.source = 'debt_entry' AND c.source_row_id = e.entry_id
                         AND c.tx_type = 'debt_signed')
                 ORDER BY e.entry_id ASC
                 LIMIT ?1",
            )?;
            let rows = stmt.query_map(params![remaining as i64], |row| {
                let description: Option<String> = row.get(4)?;
                Ok(transaction::from_debt_signed(
                    row.get(0)?,
                    row.get(1)?,
                    &row.get::<_, String>(2)?,
                    row.get(3)?,
                    description.as_deref(),
                    &row.get::<_, String>(5)?,
                    row.get(6)?,
                    row.get(7)?,
                ))
            })?;
            for row in rows {
                out.push(row?);
            }
        }

        // --- debt_entries: paid ---
        // A second, separate append rather than an edit of the signed
        // anchor. The debt's own business logic stays entirely in
        // debt_ledger.py: nothing here marks anything paid or unpaid.
        let remaining = budget.saturating_sub(out.len());
        if remaining > 0 {
            let mut stmt = conn.prepare(
                "SELECT e.entry_id, e.chat_id, e.debtor_name, e.amount_satang,
                        e.paid_by, e.paid_at
                 FROM debt_entries e
                 WHERE e.status = 'paid' AND e.paid_at IS NOT NULL
                   AND NOT EXISTS (
                       SELECT 1 FROM chain_transactions c
                       WHERE c.source = 'debt_entry' AND c.source_row_id = e.entry_id
                         AND c.tx_type = 'debt_paid')
                 ORDER BY e.entry_id ASC
                 LIMIT ?1",
            )?;
            let rows = stmt.query_map(params![remaining as i64], |row| {
                Ok(transaction::from_debt_paid(
                    row.get(0)?,
                    row.get(1)?,
                    &row.get::<_, String>(2)?,
                    row.get(3)?,
                    row.get(4)?,
                    row.get(5)?,
                ))
            })?;
            for row in rows {
                out.push(row?);
            }
        }
    }

    // --- expenses ---
    // Expense rows are editable and soft-deletable, so one anchor per
    // distinct `updated_at`: the first is expense_added, later ones are
    // expense_revised, and the soft delete is expense_deleted. The
    // `c.occurred_at >= e.updated_at` guard is the anti-join.
    //
    // Known, accepted limitation: two edits inside the same clock second
    // share one `updated_at` and therefore collapse into a single anchor.
    // Recording an expense still never touches a wallet balance.
    if storage::table_exists(conn, "expenses")? {
        let remaining = budget.saturating_sub(out.len());
        if remaining > 0 {
            let mut stmt = conn.prepare(
                "SELECT e.expense_id, e.chat_id, e.user_id, e.amount_satang, e.category,
                        e.description, e.expense_date, e.wallet_transaction_id,
                        e.debt_entry_id, e.created_at, e.updated_at, e.deleted_at
                 FROM expenses e
                 WHERE NOT EXISTS (
                       SELECT 1 FROM chain_transactions c
                       WHERE c.source = 'expense' AND c.source_row_id = e.expense_id
                         AND c.occurred_at >= e.updated_at)
                 ORDER BY e.expense_id ASC
                 LIMIT ?1",
            )?;
            let rows = stmt.query_map(params![remaining as i64], |row| {
                let description: Option<String> = row.get(5)?;
                let created_at: i64 = row.get(9)?;
                let updated_at: i64 = row.get(10)?;
                let deleted_at: Option<i64> = row.get(11)?;
                let kind = expense_kind(created_at, updated_at, deleted_at);
                Ok(transaction::from_expense_row(
                    kind,
                    row.get(0)?,
                    row.get(1)?,
                    row.get(2)?,
                    row.get(3)?,
                    &row.get::<_, String>(4)?,
                    description.as_deref(),
                    &row.get::<_, String>(6)?,
                    row.get(7)?,
                    row.get(8)?,
                    updated_at,
                    updated_at,
                ))
            })?;
            for row in rows {
                out.push(row?);
            }
        }
    }

    Ok(out)
}

/// Which expense anchor a row's current state represents.
pub fn expense_kind(created_at: i64, updated_at: i64, deleted_at: Option<i64>) -> &'static str {
    if deleted_at.is_some() {
        TX_EXPENSE_DELETED
    } else if updated_at > created_at {
        TX_EXPENSE_REVISED
    } else {
        TX_EXPENSE_ADDED
    }
}

/// Appends at most one block containing everything not yet anchored.
///
/// The candidate read and the block insert happen inside the same
/// `BEGIN IMMEDIATE` transaction, so a concurrent anchor run cannot pick up
/// the same rows: the second one blocks on the write lock and then finds the
/// anti-join empty. If the process dies mid-run, nothing is committed and
/// the next run simply re-collects the same rows.
pub fn anchor(conn: &mut Connection, max_tx: usize) -> Result<serde_json::Value> {
    storage::init_schema(conn)?;
    let budget = if max_tx == 0 { DEFAULT_MAX_TX_PER_BLOCK } else { max_tx };

    let tx = conn.transaction_with_behavior(TransactionBehavior::Immediate)?;
    let result = {
        let tip_height: i64 = {
            use rusqlite::OptionalExtension;
            tx.query_row("SELECT MAX(height) FROM chain_blocks", [], |row| {
                row.get::<_, Option<i64>>(0)
            })
            .optional()?
            .flatten()
            .ok_or_else(|| {
                ChainError::Invalid("chain is not initialized; run `init` first".to_string())
            })?
        };
        let previous_hash: String = tx.query_row(
            "SELECT hash FROM chain_blocks WHERE height = ?1",
            params![tip_height],
            |row| row.get(0),
        )?;

        let pending = collect_pending(&tx, budget)?;
        if pending.is_empty() {
            serde_json::json!({
                "ok": true,
                "anchored": 0,
                "height": tip_height,
                "block": serde_json::Value::Null,
            })
        } else {
            let block = Block::new(tip_height + 1, previous_hash, now_epoch(), pending);
            storage::insert_block(&tx, &block, now_epoch())?;
            tx.execute(
                "INSERT INTO chain_meta (key, value) VALUES ('last_anchor_at', ?1)
                 ON CONFLICT(key) DO UPDATE SET value = excluded.value",
                params![now_epoch().to_string()],
            )?;
            serde_json::json!({
                "ok": true,
                "anchored": block.tx_count,
                "height": block.height,
                "block": block.to_json(false),
            })
        }
    };
    tx.commit()?;
    Ok(result)
}

// ---------------- Read operations ----------------

pub fn status(conn: &Connection) -> Result<serde_json::Value> {
    storage::init_schema(conn)?;
    let height = storage::height(conn)?;
    let tx_count = storage::transaction_count(conn)?;
    let tip_hash = match height {
        Some(h) => storage::load_block(conn, h)?.map(|b| b.hash),
        None => None,
    };
    let pending = collect_pending(conn, DEFAULT_MAX_TX_PER_BLOCK)?.len() as i64;
    Ok(serde_json::json!({
        "ok": true,
        "initialized": height.is_some(),
        "height": height,
        "block_count": height.map(|h| h + 1),
        "transaction_count": tx_count,
        "tip_hash": tip_hash,
        "pending_transactions": pending,
        "pending_capped_at": DEFAULT_MAX_TX_PER_BLOCK,
        "last_anchor_at": storage::get_meta(conn, "last_anchor_at")?
            .and_then(|v| v.parse::<i64>().ok()),
        "chain_domain": crate::CHAIN_DOMAIN,
    }))
}

pub fn get_block(conn: &Connection, height: i64) -> Result<serde_json::Value> {
    let mut block = match storage::load_block(conn, height)? {
        Some(b) => b,
        None => {
            return Ok(serde_json::json!({"ok": false, "error": "BLOCK_NOT_FOUND",
                                         "height": height}))
        }
    };
    block.transactions = storage::load_block_transactions(conn, height)?;
    Ok(serde_json::json!({"ok": true, "block": block.to_json(true)}))
}

/// Looks up by exact `tx_ref` first, then falls back to every anchor
/// recorded for a `source:row_id` pair.
pub fn find_transaction(conn: &Connection, needle: &str) -> Result<serde_json::Value> {
    if let Some((tx, height)) = storage::find_transaction_by_ref(conn, needle)? {
        let mut value = tx.to_json();
        value["block_height"] = serde_json::json!(height);
        value["hash_consistent"] = serde_json::json!(tx.hash_is_consistent());
        return Ok(serde_json::json!({"ok": true, "matches": [value]}));
    }

    let (source, row_id) = match parse_source_ref(needle) {
        Some(pair) => pair,
        None => {
            return Ok(serde_json::json!({"ok": false, "error": "TRANSACTION_NOT_FOUND",
                                         "query": needle}))
        }
    };
    let found = storage::find_transactions_by_source(conn, source, row_id)?;
    if found.is_empty() {
        return Ok(serde_json::json!({"ok": false, "error": "TRANSACTION_NOT_FOUND",
                                     "query": needle}));
    }
    let matches: Vec<serde_json::Value> = found
        .into_iter()
        .map(|(tx, height)| {
            let mut value = tx.to_json();
            value["block_height"] = serde_json::json!(height);
            value["hash_consistent"] = serde_json::json!(tx.hash_is_consistent());
            value
        })
        .collect();
    Ok(serde_json::json!({"ok": true, "matches": matches}))
}

/// Accepts `wallet_tx:12`, `debt_entry:7`, `expense:3`, or a bare number
/// (treated as a wallet transaction id, since that is the id users see in
/// /history and /wallet_admin transactions).
fn parse_source_ref(needle: &str) -> Option<(&'static str, i64)> {
    if let Ok(id) = needle.trim().parse::<i64>() {
        return Some((SOURCE_WALLET, id));
    }
    let (prefix, rest) = needle.split_once(':')?;
    let id = rest.split(':').next()?.parse::<i64>().ok()?;
    match prefix {
        SOURCE_WALLET | "wallet" => Some((SOURCE_WALLET, id)),
        SOURCE_DEBT | "debt" => Some((SOURCE_DEBT, id)),
        SOURCE_EXPENSE => Some((SOURCE_EXPENSE, id)),
        _ => None,
    }
}

// ---------------- Reconciliation against the live source rows ----------------

/// Recomputes each anchored payload from the source row as it stands today
/// and reports any divergence.
///
/// `verify` proves the chain has not been rewritten. This proves the
/// *database* has not been edited behind the chain's back — an UPDATE
/// straight against `wallet_transactions` shows up here even though the
/// chain itself is perfectly well formed.
///
/// Rows with anchoring still pending are excluded rather than reported, so
/// a normal backlog never looks like tampering.
fn push_mismatch(
    problems: &mut Vec<serde_json::Value>,
    tx_ref: &str,
    stored: &str,
    recomputed: &str,
) {
    if stored != recomputed {
        problems.push(serde_json::json!({
            "tx_ref": tx_ref,
            "stored_payload_hash": stored,
            "recomputed_payload_hash": recomputed,
        }));
    }
}

pub fn reconcile(conn: &Connection, limit: usize) -> Result<serde_json::Value> {
    let cap = if limit == 0 { 5000 } else { limit } as i64;
    let mut checked: i64 = 0;
    let mut problems: Vec<serde_json::Value> = Vec::new();


    if storage::table_exists(conn, "wallet_transactions")? {
        let mut stmt = conn.prepare(
            "SELECT c.payload_hash, w.transaction_id, w.chat_id, w.user_id, w.type,
                    w.amount_satang, w.balance_before_satang, w.balance_after_satang,
                    w.status, w.reference_id, w.counterparty_user_id, w.created_at
             FROM chain_transactions c
             JOIN wallet_transactions w ON w.transaction_id = c.source_row_id
             WHERE c.source = 'wallet_tx'
             ORDER BY c.tx_id ASC LIMIT ?1",
        )?;
        let rows = stmt.query_map(params![cap], |row| {
            let stored: String = row.get(0)?;
            let rebuilt = transaction::from_wallet_row(
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                &row.get::<_, String>(4)?,
                row.get(5)?,
                row.get(6)?,
                row.get(7)?,
                &row.get::<_, String>(8)?,
                row.get(9)?,
                row.get(10)?,
                row.get(11)?,
            );
            Ok((stored, rebuilt))
        })?;
        for row in rows {
            let (stored, rebuilt) = row?;
            checked += 1;
            push_mismatch(&mut problems, &rebuilt.tx_ref, &stored, &rebuilt.payload_hash);
        }
    }

    if storage::table_exists(conn, "debt_entries")? {
        let mut stmt = conn.prepare(
            "SELECT c.payload_hash, c.tx_type, e.entry_id, e.chat_id, e.debtor_name,
                    e.amount_satang, e.item_description, e.entry_date, e.recorded_by,
                    e.created_at, e.paid_by, e.paid_at
             FROM chain_transactions c
             JOIN debt_entries e ON e.entry_id = c.source_row_id
             WHERE c.source = 'debt_entry'
             ORDER BY c.tx_id ASC LIMIT ?1",
        )?;
        let rows = stmt.query_map(params![cap], |row| {
            let stored: String = row.get(0)?;
            let tx_type: String = row.get(1)?;
            let description: Option<String> = row.get(6)?;
            let paid_at: Option<i64> = row.get(11)?;
            let rebuilt = if tx_type == TX_DEBT_PAID {
                transaction::from_debt_paid(
                    row.get(2)?,
                    row.get(3)?,
                    &row.get::<_, String>(4)?,
                    row.get(5)?,
                    row.get(10)?,
                    paid_at.unwrap_or(0),
                )
            } else {
                transaction::from_debt_signed(
                    row.get(2)?,
                    row.get(3)?,
                    &row.get::<_, String>(4)?,
                    row.get(5)?,
                    description.as_deref(),
                    &row.get::<_, String>(7)?,
                    row.get(8)?,
                    row.get(9)?,
                )
            };
            Ok((stored, rebuilt))
        })?;
        for row in rows {
            let (stored, rebuilt) = row?;
            checked += 1;
            push_mismatch(&mut problems, &rebuilt.tx_ref, &stored, &rebuilt.payload_hash);
        }
    }

    if storage::table_exists(conn, "expenses")? {
        let mut stmt = conn.prepare(
            "SELECT c.payload_hash, e.expense_id, e.chat_id, e.user_id, e.amount_satang,
                    e.category, e.description, e.expense_date, e.wallet_transaction_id,
                    e.debt_entry_id, e.created_at, e.updated_at, e.deleted_at
             FROM chain_transactions c
             JOIN expenses e ON e.expense_id = c.source_row_id
             WHERE c.source = 'expense' AND c.occurred_at >= e.updated_at
             ORDER BY c.tx_id ASC LIMIT ?1",
        )?;
        let rows = stmt.query_map(params![cap], |row| {
            let stored: String = row.get(0)?;
            let description: Option<String> = row.get(6)?;
            let created_at: i64 = row.get(10)?;
            let updated_at: i64 = row.get(11)?;
            let deleted_at: Option<i64> = row.get(12)?;
            let kind = expense_kind(created_at, updated_at, deleted_at);
            let rebuilt = transaction::from_expense_row(
                kind,
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                row.get(4)?,
                &row.get::<_, String>(5)?,
                description.as_deref(),
                &row.get::<_, String>(7)?,
                row.get(8)?,
                row.get(9)?,
                updated_at,
                updated_at,
            );
            Ok((stored, rebuilt))
        })?;
        for row in rows {
            let (stored, rebuilt) = row?;
            checked += 1;
            push_mismatch(&mut problems, &rebuilt.tx_ref, &stored, &rebuilt.payload_hash);
        }
    }

    let mismatches = problems.len();
    let clean = problems.is_empty();
    let shown: Vec<serde_json::Value> = problems.into_iter().take(50).collect();
    Ok(serde_json::json!({
        "ok": clean,
        "checked": checked,
        "mismatches": mismatches,
        "problems": shown,
    }))
}

/// Convenience wrapper used by the CLI and by the integration tests.
pub fn verify(conn: &Connection, from: i64) -> Result<serde_json::Value> {
    let report = crate::validation::validate_chain(conn, from)?;
    Ok(report.to_json())
}

/// Guard used by the CLI: a genesis-only chain is valid but has nothing to
/// anchor, and callers should be able to tell the two apart.
pub fn is_initialized(conn: &Connection) -> Result<bool> {
    Ok(storage::height(conn)?.is_some())
}
